code/__HELPERS/sanitize_values.dm

```diff
@@ -51,40 +51,49 @@
 				return default
 	return default
 
-/proc/sanitize_hexcolor(color, desired_format=3, include_crunch=0, default)
+/proc/sanitize_hexcolor(color, desired_format = 3, include_crunch = FALSE, default)
 	var/crunch = include_crunch ? "#" : ""
 	if(!istext(color))
 		color = ""
 
 	var/start = 1 + (text2ascii(color, 1) == 35)
 	var/len = length(color)
 	var/char = ""
-	// RRGGBB -> RGB but awful
-	var/convert_to_shorthand = desired_format == 3 && length_char(color) > 3
+	// Used for conversion between RGBA hex formats.
+	var/format_input_ratio = len ? desired_format / (length_char(color)-(start-1)) : -1
 
 	. = ""
 	var/i = start
 	while(i <= len)
 		char = color[i]
+		i += length(char)
 		switch(text2ascii(char))
 			if(48 to 57)		//numbers 0 to 9
 				. += char
 			if(97 to 102)		//letters a to f
 				. += char
 			if(65 to 70)		//letters A to F
-				. += lowertext(char)
+				char = lowertext(char)
+				. += char
 			else
 				break
-		i += length(char)
-		if(convert_to_shorthand && i <= len) //skip next one
-			i += length(color[i])
-
-	if(length_char(.) != desired_format)
-		if(default)
-			return default
-		return crunch + repeat_string(desired_format, "0")
+		switch(format_input_ratio)
+			if(0 to 4/6) //skip next one. RRGGBB(AA) -> RGB(A)
+				i += length(color[i])
+			if(6/4 to INFINITY) //Add current char again. RGB(A) -> RRGGBB(AA)
+				. += char
 
-	return crunch + .
+	if(length_char(.) == desired_format)
+		return crunch + .
+	switch(format_input_ratio) //Add or remove alpha channel depending on desired format.
+		if(3/8, 3/4, 6/4)
+			return copytext(., 1, desired_format+1)
+		if(4/6)
+			return . + "f"
+		if(4/3, 8/3)
+			return . + ((desired_format == 4) ? "f" : "ff")
```

Comment:
This doesn't look right to me, comparing floats directly is pretty dangerous.

Reply:
I don't think this is resolved at all, comparing floats is dangerous even if you write out ".25". I'd rather you just check the numerator and denominator specifically, since that's more safe to compare.